kore/src/Kore/Internal/Predicate.hs

```diff
@@ -119,438 +184,642 @@ import Pretty
-pattern PredicateTrue  <- GenericPredicate (Recursive.project -> _ :< TopF _)
+instance Monoid HasChanged where
+    mempty = NotChanged
 
-pattern PredicateAnd
-    :: Predicate variable -> Predicate variable -> Predicate variable
-pattern PredicateAnd p1 p2 <-
-    GenericPredicate (And_ _ (GenericPredicate -> p1) (GenericPredicate -> p2))
 {-|'isFalse' checks whether a predicate is obviously bottom.
 -}
-isFalse :: TopBottom patt => GenericPredicate patt -> Bool
+isFalse :: TopBottom patt => patt -> Bool
 isFalse = isBottom
 
-binaryOperator
-    :: (HasCallStack, InternalVariable variable)
-    => (TermLike variable -> TermLike  variable -> TermLike  variable)
-    -> Predicate variable -> Predicate variable -> Predicate variable
-binaryOperator operator (GenericPredicate first) (GenericPredicate second) =
-    GenericPredicate (operator first sortedSecond)
-  where
-    sortedSecond = TermLike.fullyOverrideSort (termLikeSort first) second
+isTrue :: TopBottom patt => patt -> Bool
+isTrue = not . isBottom
 
-{-| 'makeMultipleAndPredicate' combines a list of Predicates with 'and',
-doing some simplification.
--}
+makeTruePredicate' ::
+    InternalVariable variable
+    => (Predicate variable, HasChanged)
+makeTruePredicate' = (synthesize (TopF Top {topSort = ()})
+    , NotChanged)
 
-makeMultipleAndPredicate
-    :: (HasCallStack, InternalVariable variable)
-    => [Predicate variable]
-    -> Predicate variable
-makeMultipleAndPredicate =
-    foldl' makeAndPredicate makeTruePredicate_ . nubOrd
-    -- 'and' is idempotent so we eliminate duplicates
+makeTruePredicate :: InternalVariable variable => Predicate variable
+makeTruePredicate = fst makeTruePredicate'
 
-{- | Flatten a 'Predicate' with 'And' at the top.
-'getMultiAndPredicate' is the inverse of 'makeMultipleAndPredicate', up to
-associativity.
- -}
+makeFalsePredicate' ::
+    InternalVariable variable
+    => (Predicate variable, HasChanged)
+makeFalsePredicate' = (synthesize $ BottomF Bottom {bottomSort = ()}
+    , NotChanged)
 
-getMultiAndPredicate
-    :: Predicate variable
-    -> [Predicate variable]
-getMultiAndPredicate original@(GenericPredicate termLike) =
-    case termLike of
-        And_ _ left right ->
-            concatMap (getMultiAndPredicate . GenericPredicate) [left, right]
-        _ -> [original]
-
-{-| 'makeMultipleOrPredicate' combines a list of Predicates with 'or',
-doing some simplification.
--}
-makeMultipleOrPredicate
+makeFalsePredicate :: InternalVariable variable => Predicate variable
+makeFalsePredicate = fst makeFalsePredicate'
+
+makeNotPredicate'
     :: InternalVariable variable
-    => [Predicate variable]
+    => Predicate variable
+    -> (Predicate variable, HasChanged)
+makeNotPredicate' p
+  | isTop p = (makeFalsePredicate, Changed)
+  | isBottom p = (makeTruePredicate, Changed)
+  | otherwise = (synthesize $ NotF Not
+    { notSort = ()
+    , notChild = p
+    }
+    , NotChanged)
```

Comment:
When splitting a tuple over multiple lines, our style guide calls for aligning the brackets and commas:

```suggestion
  | otherwise =
    ( synthesize $ NotF Not
        { notSort = ()
        , notChild = p
        }
    , NotChanged
    )
```